test/unit/logic/delegate.js

```diff
@@ -0,0 +1,535 @@
+		it('should update username to null if account did not have a username before', function (done) {
+			delete sender.username;
+			sender.nameexist = 0;
+
+			accountsMock.setAccountAndGet.once().withExactArgs({
+				address: sender.address,
+				u_isDelegate: 0,
+				isDelegate: 0,
+				username: null,
+				u_username: null
+			}, dummyCb);
+
+			delegate.undoUnconfirmed(trs, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+
+		it('should call accounts.setAccountAndGet module with correct parameters', function (done) {
+			accountsMock.setAccountAndGet.once().withExactArgs({
+				address: sender.address,
+				u_isDelegate: 0,
+				isDelegate: 0,
+				username: null,
+				u_username: null
+			}, dummyCb);
+
+			delegate.undoUnconfirmed(trs, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+	});
+
+	describe('objectNormalize', function () {
+
+		var schemaDynamicTest = new SchemaDynamicTest({
+			testStyle: SchemaDynamicTest.TEST_STYLE.THROWABLE,
+			customPropertyAssertion: function (input, expectedType, property, err) {
+				expect(err).to.equal('Failed to validate delegate schema: Expected type ' + expectedType + ' but found type ' + input.expectation);
+			}
+		});
+
+		after(function () {
+			describe('schema dynamic tests', function () {
+				schemaDynamicTest.schema.shouldFailAgainst.nonObject.property(delegate.objectNormalize, trs, 'asset.delegate');
+
+				schemaDynamicTest.schema.shouldFailAgainst.nonString.property(delegate.objectNormalize, trs, 'asset.delegate.username');
+			});
+		});
+
+		it('should use the correct format to validate against', function () {
+			var library = Delegate.__get__('library');
+			var schemaSpy = sinon.spy(library.schema, 'validate');
+			delegate.objectNormalize(trs);
+			expect(schemaSpy.calledOnce).to.equal(true);
+			expect(schemaSpy.calledWithExactly(trs.asset.delegate, Delegate.prototype.schema)).to.equal(true);
+			schemaSpy.restore();
+		});
+
+		it('should throw error when asset schema is invalid', function () {
+			trs.asset.delegate.username = '';
+
+			expect(function () {
+				delegate.objectNormalize(trs);
+			}).to.throw('Failed to validate delegate schema: Object didn\'t pass validation for format username: ');
+		});
+
+		it('should return transaction when asset is valid', function () {
+			expect(delegate.objectNormalize(trs)).to.eql(trs);
+		});
+	});
+
+	describe.skip('dbRead', function () {
+
+		it('should return null when username is not set ', function () {
+			delete rawTrs.d_username;
+
+			expect(delegate.dbRead(rawTrs)).to.eql(null);
+		});
+
+		it('should be okay for valid input', function () {
+			var expectedAsset = {
+				address: sender.address,
+				publicKey: sender.publicKey,
+				username: trs.asset.delegate.username
+			};
+
+			expect(delegate.dbRead(rawTrs).delegate).to.eql(expectedAsset);
+		});
+	});
+
+	describe.skip('dbSave', function () {
+
+		it('should be okay for valid input', function () {
```

Comment:
please add separate test to check if each field is being returned

Reply:
it deeply checks for all object properties. Besides, this is a promiseObject, and we expect all the properties to be present in it together.